atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/limiter/ClientLimiterImpl.java

```diff
@@ -0,0 +1,48 @@
+/*
+ * (c) Copyright 2024 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.keyvalue.cassandra.limiter;
+
+import com.palantir.atlasdb.keyvalue.cassandra.limiter.SafeSemaphore.CloseablePermit;
+import com.palantir.atlasdb.util.MetricsManager;
+import java.io.IOException;
+import java.util.Map;
+import org.apache.cassandra.thrift.TimedOutException;
+import org.apache.thrift.TException;
+
+public class ClientLimiterImpl implements ClientLimiter {
+
+    private final SafeSemaphore concurrentRangeScans;
+
+    public ClientLimiterImpl(int maxConcurrentRangeScans, MetricsManager metricsManager) {
+        concurrentRangeScans = SafeSemaphore.of(maxConcurrentRangeScans);
+
+        metricsManager.registerOrGet(
+                ClientLimiter.class,
+                "concurrent-range-scans",
+                () -> maxConcurrentRangeScans - concurrentRangeScans.availablePermits(),
+                Map.of());
+    }
+
+    @Override
+    public <T> T limitGetRangeSlices(ThriftExceptionThrowingSupplier<T> closure) throws TException {
+        try (CloseablePermit permit = concurrentRangeScans.tryAcquire().orElseThrow(TimedOutException::new)) {
```

Comment:
@ellisjoe why do we want to throw a non-retryable exception? I'd expect to throw a QosException so that the clients can retry after some duration.

Reply:
I guess the end goal in my mind is to get out of a world where these range scans are happening at all. Given that, I'm more comfortable hard failing here (the alternative is that we end up hard failing at some unknown time in the future when a cluster is overloaded).

Reply:
that's fair - this could be disruptive until we provide an alternative storage primitive like a low-lift search functionality to the clients or they figure out workarounds for it. 
It's fine to proceed though as we've bumped the limit quite a bit.